// @microsoft/ads-service-downloader/out/logger.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the Source EULA. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleLogger = exports.Logger = void 0;
class Logger {
    constructor(eventEmitter) {
        this.eventEmitter = eventEmitter;
    }
    verbose(message) {
        this.log(0 /* LogLevel.Verbose */, message);
    }
    info(message) {
        this.log(1 /* LogLevel.Information */, message);
    }
    warn(message) {
        this.log(2 /* LogLevel.Warning */, message);
    }
    error(message) {
        this.log(3 /* LogLevel.Error */, message);
    }
    critical(message) {
        this.log(4 /* LogLevel.Critical */, message);
    }
    log(level, message) {
        this.eventEmitter.emit("log_emitted" /* Events.LOG_EMITTED */, level, message);
    }
}
exports.Logger = Logger;
class ConsoleLogger {
    verbose(message) {
        console.log(message);
    }
    info(message) {
        console.info(message);
    }
    warn(message) {
        console.warn(message);
    }
    error(message) {
        console.error(message);
    }
    critical(message) {
        console.error(message);
    }
}
exports.ConsoleLogger = ConsoleLogger;
